File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if(!s) return 0;
    let res = 1;
    if(s.length === 1) return 1;
    let i = 0, j = 1;
    let dict = {};
    dict[s[0]] = 0;
    while(j < s.length) {
        if(dict[s[j]] !== undefined) {
            if(j-i > res) res = j-i;
            if(dict[s[j]] + 1 > i) i = dict[s[j]]+1;
            dict[s[j]] = j;
        } else
            dict[s[j]] = j;
        j++;
    }
    return j-i > res? j-i : res;
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if(not s): return 0
        i = 0
        j = 1
        length = 1
        dict = {}
        dict[s[0]] = 0
        while(j < len(s)):
            if(s[j] in dict):
                if(dict[s[j]] >= i):
                    length = j-i if j-i > length else length
                    i = dict[s[j]]+1
            dict[s[j]] = j
            j += 1
        return length if length > j-i else j-i